Cleanup hasOverloads

9rnsr opened the pull request:
Remove `Dsymbol.hasOverloads()` because it's confusing method.

9rnsr commented:
This front-end cleanup is necessary for:

https://trello.com/c/ihKhGeiu/115-reg2-031-forward-reference-to-inferred-return-type-of-function-call-when-using-auto-return-type

yebblies commented on src/expression.d:
```diff
@@ -3676,6 +3676,8 @@ public:
 }
 
 /***********************************************************
+ * Don't be generated by parser.
```
`"Won't be generated"` or `"Isn't generated"`

yebblies commented on src/expression.d:
```diff
@@ -3695,6 +3697,13 @@ public:
     }
 
+    /****************************************
+     * Resolve a symbol `s` to correcponding concrete Expression object.
```
`corresponding`

I think a better description is that it resolves a symbol and wraps it in an expression object.

9rnsr replied:
Done.

yebblies commented on src/mtype.d:
```diff
@@ -7979,8 +7979,9 @@ public:
             e = e.semantic(sc);
             return e;
         }
-        auto de = new DotVarExp(e.loc, e, d);
```
Re-using `e` is not an improvement.

9rnsr replied:
It's improvement for print debugging.

9rnsr replied:
`de` is DotVarExp, so we cannot assign the `de.semantic` result to it. So, if we want to print the `de.semantic` result, the original code should be rewirten to:

```d
        auto de = new DotVarExp(e.loc, e, d, d.hasOverloads());
-       return de.semantic(sc);
+      e = de.semantic(sc);
+      printf("%s", e.toChars());
+      return e;
```

By reusing `Expression e` variable, now we can insert one `printf` line.

```d
        e = new DotVarExp(e.loc, e, d, d.isFuncDeclaration() !is null);
        e = e.semantic(sc);
+      printf("%s", e.toChars());
        return e;
```

ibuclaw replied:
Personally I'd call ``de.toChars()`` from a debugger.  But I don't think that this is blocking anything.

9rnsr commented on src/expression.d:
```diff
@@ -6027,7 +6036,7 @@ public:
         if (VarDeclaration vd = var.isVarDeclaration())
         {
-            hasOverloads = 0;
+            hasOverloads = false;//assert(!hasOverloads);
```
It was a junk. Removed.

9rnsr commented:
Ping to @yebblies and @klickverbot .

In `TypeStruct` (or `TypeClass`) `dotExp()`, it resolves a form `e.ident` to an actual expression. And if `ident` is a function, it's not yet known whether it has overloads or not. In other words, following semantic analysis should assume that the function would have some overloads. That's　why we should set `DotVarExp.hasOverloads` to `true` if `d` is a `FuncDeclaration`.

The logic will also fit to `TypeExp.semantic` and `ScopeExp.semantic`.

9rnsr committed:
Fix hasOverloads field settings to correctly handle overloaded function names

Grep-ed `new (Dsymbol|Var|DotVar|SymOff|Delegate)Exp` and has checked them all.

Usually, when a function symbol appears in expression, e.g. `func.d()`, `&func` etc, the `func` has overload candidates. So, usually `hasOverloads` field should be set to `true` until it's resolved. Changing their default arguments to `true`, and supply `false` explicitly around internal code generations.

9rnsr commented:
Updated to fix all settings of `(Dsymbol|Var|DotVar|SymOff|Delegate)Exp.hasOverloads` fields.

Usually, when a function symbol appears in expression, e.g. `func.d()`, `&func` etc, the `func` has overload candidates. So, usually `hasOverloads` field should be set to `true` until it's resolved. Changing their default arguments to `true`, and supply `false` explicitly around internal code generations.

Now this PR is shared by #2130 and #5202.

ibuclaw commented on src/expression.d:
```diff
@@ -3732,7 +3734,7 @@ public:
     bool hasOverloads;
 
-    extern (D) this(Loc loc, Dsymbol s, bool hasOverloads = false)
+    extern (D) this(Loc loc, Dsymbol s, bool hasOverloads = true)
```
Maybe it's better just to force hasOverloads to be explicitly passed?

9rnsr replied:
I don't think so. Half of `hasOverloads` would need to be set to `false` for `VarDeclaration`s. For them, explicit argument passing is just verbose. The default argument `true` works for good default behavior.

ibuclaw commented:
Auto-merge toggled on

ibuclaw commented:
Ok, well on the whole I have nothing against this.

9rnsr commented:
Thanks!
